test/test_images/probe_helper/probe/helper.go

```diff
@@ -0,0 +1,374 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package probe
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/storage"
+	cepubsub "github.com/cloudevents/sdk-go/protocol/pubsub/v2"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
+	"go.uber.org/zap"
+
+	"knative.dev/pkg/logging"
+
+	schemasv1 "github.com/google/knative-gcp/pkg/schemas/v1"
+	"github.com/google/knative-gcp/test/test_images/probe_helper/utils"
+)
+
+const (
+	probeEventTimeoutExtension     = "timeout"
+	probeEventRequestHostExtension = "requesthost"
+	probeEventRequestHostHeader    = "Ce-RequestHost"
```

Comment:
I think we need to better document this. We are relying on the host name of the received event to identify which trigger delivered the event. I think something like `targetTrigger` is easier to understand. Then the receiver can translate the host into a trigger, assuming there is a 1:1 mapping between a trigger type and host (1 trigger per type in a namespace).

And this only applies to the AuditLogSource. Maybe we can move this logic to the AudigLogSourceHandler?

Reply:
I added some documentation about this, and changed to a targetpath instead of service or trigger:
```golang
	// This is the CloudEvent extension which holds the target path of probe requests.
	// The path of receiver probe requests is injected into this extension by the
	// receiver client. The corresponding forward probe requests should include
	// this extension with the same path. For example:
	//
	// If a source is configured to sink events to:
	//   http://probe-helper-receiver.cloud-run-events-probe.svc.cluster.local/some-path-goes-here
	//
	// then, each event generated by the source and passed to the probe helper
	// receiver client will include the header 'Ce-Targetpath: /some-path-goes-here'.
	// And in order for the probe to succeed, the corresponding forward probe should
	// also include the header 'Ce-Targetpath: /some-path-goes-here'.
	probeEventTargetPathExtension = "targetpath"
```

Reply:
> each event generated by the source and passed to the probe helper receiver client will include the header 'Ce-Targetpath: /some-path-goes-here'.

We don't have control on the event headers generated by source, do we? What we can control is by pointing the source/trigger to this specific path in the receiver uri.

Reply:
Correct. I inject the path into the header of received CloudEvents with receiver client middleware: https://github.com/google/knative-gcp/blob/342f0e314558b88723e04ee0e81baaf088cb670d/test/test_images/probe_helper/probe/helper.go#L308-L315

This seemed like the simplest way to access the request path while still getting to use a CloudEvent client.

Reply:
Sorry I missed that part. Two follow ups:
1. The code only sets the header if it's not empty, which I think is wrong.  I think the intention here is to **always** set the header to be the receiver path.
2. If 1 is correct, please document this behavior. Also I think it's probably cleaner to introduce a new header, something like "probeEventReceiverPathHeader" to differentiate from the initial targetPath

Reply:
Done.